apollo-router/src/plugins/coprocessor/execution.rs

```diff
@@ -0,0 +1,1011 @@
+    /// Send the headers
+    pub(super) headers: bool,
+    /// Send the context
+    pub(super) context: bool,
+    /// Send the body
+    pub(super) body: bool,
+    /// Send the SDL
+    pub(super) sdl: bool,
+    /// Send the method
+    pub(super) method: bool,
+    /// Send the query plan
+    pub(super) query_plan: bool,
+}
+
+/// What information is passed to a router request/response stage
+#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize, JsonSchema)]
+#[serde(default, deny_unknown_fields)]
+pub(super) struct ExecutionResponseConf {
+    /// Send the headers
+    pub(super) headers: bool,
+    /// Send the context
+    pub(super) context: bool,
+    /// Send the body
+    pub(super) body: bool,
+    /// Send the SDL
+    pub(super) sdl: bool,
+    /// Send the HTTP status
+    pub(super) status_code: bool,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize, JsonSchema)]
+#[serde(default)]
+pub(super) struct ExecutionStage {
+    /// The request configuration
+    pub(super) request: ExecutionRequestConf,
+    // /// The response configuration
+    pub(super) response: ExecutionResponseConf,
+}
+
+impl ExecutionStage {
+    pub(crate) fn as_service<C>(
+        &self,
+        http_client: C,
+        service: execution::BoxService,
+        coprocessor_url: String,
+        sdl: Arc<String>,
+    ) -> execution::BoxService
+    where
+        C: Service<hyper::Request<Body>, Response = hyper::Response<Body>, Error = BoxError>
+            + Clone
+            + Send
+            + Sync
+            + 'static,
+        <C as tower::Service<http::Request<Body>>>::Future: Send + 'static,
+    {
+        let request_layer = (self.request != Default::default()).then_some({
+            let request_config = self.request.clone();
+            let coprocessor_url = coprocessor_url.clone();
+            let http_client = http_client.clone();
+            let sdl = sdl.clone();
+
+            OneShotAsyncCheckpointLayer::new(move |request: execution::Request| {
+                let request_config = request_config.clone();
+                let coprocessor_url = coprocessor_url.clone();
+                let http_client = http_client.clone();
+                let sdl = sdl.clone();
+
+                async move {
+                    let mut succeeded = true;
+                    let result = process_execution_request_stage(
+                        http_client,
+                        coprocessor_url,
+                        sdl,
+                        request,
+                        request_config,
+                    )
+                    .await
+                    .map_err(|error| {
+                        succeeded = false;
+                        tracing::error!(
+                            "external extensibility: execution request stage error: {error}"
+                        );
+                        error
+                    });
+                    tracing::info!(
```

Comment:
Please use the new metrics macros rather than the tracing macros. I'm going to add a safelist to the tracing metrics macros to prevent further use within our codebase.

Reply:
In addition using the new macros will allow you to add tests for the metrics.

Reply:
Docs for new metrics macros are here: https://github.com/apollographql/router/blob/dev/dev-docs/metrics.md#using-metrics-macros

Reply:
why is the dev docs using the format `u64_counter!("test", "test description", 1, "attr" => "val");` while the macro is using `u64_counter!("test", "test description", 1, "attr" = "val");` (with `=` instead of `=>`)?
Could you check the docs again for inconsistencies?

Reply:
fixed in [40711d1](https://github.com/apollographql/router/pull/4299/commits/40711d15f38811ccde2bff23412fa8354e620339)
It would be nice to have the annotations tracing provided, like the % prefix when adding attribute values